staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go

```diff
@@ -31,6 +35,54 @@ const (
 	ToStorageLabel   = "to_storage"
 )
 
+type metricLabels struct {
+	transformationType string
+	providerName       string
+	keyIDHash          string
+}
+type metricLabelsMap struct {
+	data map[string]metricLabels
+	keys []string
+	mu   sync.RWMutex
+}
+
+func (o *metricLabelsMap) deleteFirstItem() error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if len(o.keys) == 0 {
+		return fmt.Errorf("failed to delete an empty map ")
+	}
+	delete(o.data, o.keys[0])
+	o.keys = o.keys[1:]
```

Comment:
Hmm this might be a memory leak (not sure if the underlying array will ever get smaller).

Reply:
Are you referring to `o.keys`? This call does reduce the length by 1.

Reply:
Sub slicing does not change the underlying array, it just creates another view into it.  IIUC this is why LRU caches use linked lists for tracking.